Adding `Algebra.Properties.CommutativeMonoid`

bch29 opened the pull request:
This relies on #252 and is probably the penultimate step in including all of #211.

gallais commented on src/Algebra/Properties/CommutativeMonoid.agda:
```diff
@@ -0,0 +1,168 @@
+    (head t + lookup t (suc i)) + sumₜ (rearrange (punchIn i) (tail t))   ≈⟨ +-cong (+-comm _ _) refl ⟩
+    (lookup t (suc i) + head t) + sumₜ (rearrange (punchIn i) (tail t))   ≈⟨ +-assoc _ _ _ ⟩
+    lookup t (suc i) + (head t + sumₜ (rearrange (punchIn i) (tail t)))   ∎
+
```
IMO this whole proof would become a lot clearer if you started with a set of
let-bindings:

```agda
let x = head t
    y = lookup t (suc i)
    z = sumₜ (rearrange (punchIn i) (tail t))
in begin
￼    x + sumₜ (tail t)  ≈⟨ +-cong refl (sumₜ-punchIn (tail t) i) ⟩
￼    x + (y + z)        ≈⟨ sym (+-assoc x y z) ⟩
￼    (x + y) + z        ≈⟨ +-cong (+-comm x y) refl ⟩
￼    (y + x) + z        ≈⟨ +-assoc y x z ⟩
￼    x + (y + z)   ∎

bch29 replied:
Good idea, done.

MatthewDaggitt commented on src/Algebra/Properties/CommutativeMonoid.agda:
```diff
@@ -0,0 +1,181 @@
+
+module Algebra.Properties.CommutativeMonoid {g₁ g₂} (M : CommutativeMonoid g₁ g₂) where
+
+open CommutativeMonoid M renaming (ε to 0#; _∙_ to _+_; ∙-cong to +-cong; identity to +-identity; assoc to +-assoc; comm to +-comm)
```
This seems quite key, and at the moment it's difficult to spot. Could you open this at the end of the imports, and lay each renaming instance out on a separate line?

MatthewDaggitt commented on src/Algebra/Properties/CommutativeMonoid.agda:
```diff
@@ -0,0 +1,181 @@
+sumₜ-punchIn f zero = refl
+sumₜ-punchIn {zero} t (suc ())
+sumₜ-punchIn {suc n} t (suc i) =
+  let x = head t
```
For better or worse, the usual style in the library is put these in a `where` block at the end rather than a `let`...

MatthewDaggitt commented on src/Algebra/Properties/CommutativeMonoid.agda:
```diff
@@ -0,0 +1,181 @@
+-- 'j' then selecting 'j' is the same as selecting 'i'.
+
+select-transpose : ∀ {n} t (i j : Fin n) → lookup t i ≈ lookup t j → ∀ k → (lookup (select 0# j t) ∘ PermC.transpose i j) k ≈ lookup (select 0# i t) k
+select-transpose _ i j e k with k FP.≟ i
```
You could use `...` notation with the `with` here to remove quite a lot of text in this proof.

MatthewDaggitt commented on src/Algebra/Properties/CommutativeMonoid.agda:
```diff
@@ -0,0 +1,181 @@
+
+-- The '∑' operator distributes over addition.
+
+∑-+-hom : ∀ n (f g : Fin n → Carrier) → ∑[ i < n ] f i + ∑[ i < n ] g i ≈ ∑[ i < n ] (f i + g i)
```
I think the name`∑-distrib-+` might be better for this. Perhaps(?) clearer and it follows the conventions set out in `Nat.Properties` etc.

MatthewDaggitt commented on src/Algebra/Properties/CommutativeMonoid.agda:
```diff
@@ -0,0 +1,181 @@
+
+-- '_≈_' is a congruence over 'sumTable n'.
+
+sumₜ-cong : ∀ {n} {t t′ : Table Carrier n} → t ≋ t′ → sumₜ t ≈ sumₜ t′
```
This should probably be called `sumₜ-cong-≋`

MatthewDaggitt commented on src/Algebra/Properties/CommutativeMonoid.agda:
```diff
@@ -0,0 +1,181 @@
+
+-- '_≡_' is a congruence over 'sum n'.
+
+sumₜ-cong≡ : ∀ {n} {t t′ : Table Carrier n} → t ≗ t′ → sumₜ t ≡ sumₜ t′
```
And then this should be called `sumₜ-cong-≡`

MatthewDaggitt commented:
Looks mostly good! A few additional comments.

The only thing remaining thing I'm wondering is where to put all of these. As they are proving results about the operations defined in `Algebra.Operations.CommutativeMonoid` perhaps they should live there as well.

Or perhaps the whole `Operations` subtree thing is unnecessary and they should all be moved to the `Properties` hierarchy (as we've already fiddled with the `Operations` tree in a non-backwards compatible fashion in this release...). Thoughts anyone?

bch29 commented:
I think I've addressed everything. I've also simplified a couple of proofs using the commutative monoid solver.

Regarding the question of moving things to `Algebra.Operations.CommutativeMonoid`, I prefer having definitions and proofs about them live in different files to be honest. It makes the definition file feel cleaner, and it makes it easier to make changes to the files later by improving type checking speed.

I also like the `Operations` subtree because it's clear that it is about defining composite operations in terms of the base ones, and such things don't feel like they should go in `Properties` files.

MatthewDaggitt commented:
> I think I've addressed everything. I've also simplified a couple of proofs using the commutative monoid solver.

Okay, great thanks. I'm just about highlight 3 more stylistic points. Usually wouldn't bother but as you contribute so much, thought I'd point them out for future PRs.

> Regarding the question of moving things to Algebra.Operations.CommutativeMonoid, I prefer having definitions and proofs about them live in different files to be honest. It makes the definition file feel cleaner, and it makes it easier to make changes to the files later by improving type checking speed.
>
> I also like the Operations subtree because it's clear that it is about defining composite operations in terms of the base ones, and such things don't feel like they should go in Properties files.

You're right, defining the operations in the properties file doesn't make an awful lot of sense. However at the moment we still have some proofs sitting in the `Operations` files. What we shouldn't do is split them over multiple files. Okay so perhaps we should probably move all proofs in `Operations` files into `Properties` files. One possibility would then to just have the composite operations defined directly in the algebraic record itself...

MatthewDaggitt commented on src/Algebra/Properties/CommutativeMonoid.agda:
```diff
@@ -0,0 +1,189 @@
+sumₜ-punchIn f zero = refl
+sumₜ-punchIn {zero} t (suc ())
+sumₜ-punchIn {suc n} t (suc i) =
+  begin
```
Can you align the arguments? It makes them much easier to read. If there are implicit arguments that are only used in certain cases, I usually put `{_}` when they're not needed to make it easier to parse.

Also in this particular case, you call the table `f` in the first case and then `t` in the second and third.

MatthewDaggitt commented on src/Algebra/Properties/CommutativeMonoid.agda:
```diff
@@ -0,0 +1,189 @@
+sumₜ-zero : ∀ n → sumₜ (replicate {n} 0#) ≈ 0#
+sumₜ-zero (zero) = refl
+sumₜ-zero (suc n) =
+  begin
```
We usually put `begin` on the previous line in the library, and then you can de-indent the actual proof by 2 spaces which gives you more room to play with which can be important!

bch29 replied:
What would you do with the `where` clause if `begin` is on the previous line? This looks weird to me:
```agda
sumₜ-punchIn {suc n} t (suc i) = begin
  x + sumₜ (tail t)  ≈⟨ +-cong refl (sumₜ-punchIn (tail t) i) ⟩
  x + (y + z)        ≈⟨ solve 3 (λ x y z → x ⊕ (y ⊕ z) ⊜ y ⊕ (x ⊕ z)) refl x y z ⟩
  y + (x + z)        ∎
  where
  x = head t
  y = lookup t (suc i)
  z = sumₜ (rearrange (punchIn i) (tail t))
```

and this is inconsistent with the indentation after `where` in other places:
```agda
sumₜ-punchIn {suc n} t (suc i) = begin
  x + sumₜ (tail t)  ≈⟨ +-cong refl (sumₜ-punchIn (tail t) i) ⟩
  x + (y + z)        ≈⟨ solve 3 (λ x y z → x ⊕ (y ⊕ z) ⊜ y ⊕ (x ⊕ z)) refl x y z ⟩
  y + (x + z)        ∎
  where
    x = head t
    y = lookup t (suc i)
    z = sumₜ (rearrange (punchIn i) (tail t))
```

MatthewDaggitt replied:
I don't think there's anything wrong with the first one? The `where` clearly separates the two? 

Also I think the proof would clearer if you gave `x` `y` and `z` more meaningful names, something like `t_0`, `t_i` and `\sigma[t-t_i]`?

MatthewDaggitt replied:
Also it's just occurred to me that we should define a function `remove` in `Data.Table` and prove some properties about it. That would massively clean up these proofs right?

bch29 replied:
> I don't think there's anything wrong with the first one? The where clearly separates the two.

I adopted the habit of starting begin on the next line specifically to avoid this in my own code! I just don't think text (i.e. the word `where`) is a very good visual separator on its own.

> Also it's just occurred to me that we should define a function remove in Data.Table and prove some properties about it. That would massively clean up these proofs right?

I had the same thought about defining that function, but I'm not sure what it would clean up other than making the names clearer (because it's not immediately obvious what `punchIn` does to a table).

MatthewDaggitt commented on src/Algebra/Properties/CommutativeMonoid.agda:
```diff
@@ -0,0 +1,189 @@
+select-transpose _ i j e k with k FP.≟ i
+... | yes p rewrite P.≡-≟-identity FP._≟_ {j} P.refl = sym e
+... | no ¬p with k FP.≟ j
+... | yes q rewrite proj₂ (P.≢-≟-identity FP._≟_ (¬p ∘ P.trans q ∘ P.sym)) = refl
```
After each `with` depth we usually add an additional two spaces before the `|` symbol. This indenting makes it easier to work out which level you're at.

bch29 commented:
I've also redone `sumₜ-select` so that it doesn't rely on `sum-permute`, so it should be faster to normalize now (which is something I personally care about); and the proof of `sumₜ-zero` is now a special case of `sumₜ-replicate`.

bch29 commented:
> Okay so perhaps we should probably move all proofs in Operations files into Properties files.

That sounds reasonable to me, apart from the backwards compatibility issue (fixing incompatibility now is just a matter of `import Algebra.Operations --> import Algebra.Operations.Semiring`; it would be more involved if we started moving things around).

> One possibility would then to just have the composite operations defined directly in the algebraic record itself...

I'd personally prefer to keep the basic algebraic structure definitions lean, but my opinion on that isn't overly strong.

bch29 commented:
@MatthewDaggitt you were right, defining a `remove` for `Table` made lots of things much nicer. The  `sumₜ-permute` proof is actually easy to follow now!

bch29 commented:
This new proof of `sumₜ-permute` should have faster normalization because it doesn't need to compute `↔⇒≡` any more.

MatthewDaggitt commented:
Okay glad it helped. Can you document your changes in the CHANGELOG please?

bch29 commented:
> Can you document your changes in the CHANGELOG please?

Done. I hadn't done it yet in case we ended up changing things more.

MatthewDaggitt commented:
Okay merged it in. Hold fire for a bit on the next PR as I'm going to make some tweaks.

MatthewDaggitt commented:
Actually there's quite a nasty cyclic dependency that I need time to think how to get rid of. @bch29 please do submit your final PR for your proof!
